File: resources/android/FirebaseCrashlyticsFunctions.kt
package com.codingwithrk.plugins.firebase_crashlytics

import android.os.Handler
import android.os.Looper
import androidx.fragment.app.FragmentActivity
import com.google.android.gms.tasks.Tasks
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

object FirebaseCrashlyticsFunctions {

    /** How long we're willing to block waiting on a Crashlytics Task before giving up. */
    private const val AWAIT_TIMEOUT_SECONDS = 8L

    /**
     * Records a caught exception with Crashlytics as a (non-)fatal issue.
     *
     * Parameters:
     * - message: String - Exception message
     * - className: String? - Fully qualified PHP exception class name, used to label the issue
     * - stackTrace: String? - Rendered PHP stack trace, logged as a breadcrumb before recording
     * - fatal: Boolean - Whether to flag this issue as fatal-style via a custom key
     *
     * Returns:
     * - recorded: Boolean
     */
    class RecordError(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val message = parameters["message"] as? String
                ?: return BridgeResponse.error("INVALID_PARAMETERS", "message is required")
            val className = parameters["className"] as? String ?: "PHP\\Throwable"
            val stackTrace = parameters["stackTrace"] as? String
            val fatal = parameters["fatal"] as? Boolean ?: false

            return try {
                val crashlytics = FirebaseCrashlytics.getInstance()

                if (fatal) {
                    crashlytics.setCustomKey("fatal_error", true)
                }
                stackTrace?.let { crashlytics.log("PHP stack trace:\n$it") }

                crashlytics.recordException(PhpThrowable(className, message))

                BridgeResponse.success(mapOf("recorded" to true))
            } catch (e: Exception) {
                BridgeResponse.error("RECORD_ERROR_FAILED", e.message ?: "Failed to record error")
            }
        }
    }

    /**
     * Parameters:
     * - message: String - Breadcrumb message to attach to the next report
     *
     * Returns:
     * - logged: Boolean
     */
    class Log(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val message = parameters["message"] as? String
                ?: return BridgeResponse.error("INVALID_PARAMETERS", "message is required")

            FirebaseCrashlytics.getInstance().log(message)

            return BridgeResponse.success(mapOf("logged" to true))
        }
    }

    /**
     * Parameters:
     * - identifier: String - User identifier to attach to subsequent reports
     *
     * Returns:
     * - set: Boolean
     */
    class SetUserId(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val identifier = parameters["identifier"] as? String
                ?: return BridgeResponse.error("INVALID_PARAMETERS", "identifier is required")

            FirebaseCrashlytics.getInstance().setUserId(identifier)

            return BridgeResponse.success(mapOf("set" to true))
        }
    }

    /**
     * Parameters:
     * - key: String - Custom key name
     * - value: String|Boolean|Number - Custom key value
     *
     * Returns:
     * - set: Boolean
     */
    class SetCustomKey(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val key = parameters["key"] as? String
                ?: return BridgeResponse.error("INVALID_PARAMETERS", "key is required")
            val value = parameters["value"]
                ?: return BridgeResponse.error("INVALID_PARAMETERS", "value is required")

            val crashlytics = FirebaseCrashlytics.getInstance()
            when (value) {
                is Boolean -> crashlytics.setCustomKey(key, value)
                is Int -> crashlytics.setCustomKey(key, value)
                is Long -> crashlytics.setCustomKey(key, value)
                is Float -> crashlytics.setCustomKey(key, value)
                is Double -> crashlytics.setCustomKey(key, value)
                is Number -> crashlytics.setCustomKey(key, value.toDouble())
                else -> crashlytics.setCustomKey(key, value.toString())
            }

            return BridgeResponse.success(mapOf("set" to true))
        }
    }

    /**
     * Forces an immediate, unrecoverable crash - used to verify a Crashlytics
     * integration end-to-end. The throw is posted to the next main-thread loop
     * iteration so it happens outside of the bridge dispatcher's call stack,
     * guaranteeing a real crash instead of being caught and turned into an
     * error response.
     *
     * Parameters:
     * - reason: String? - Optional crash message
     */
    class Crash(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val reason = parameters["reason"] as? String
                ?: "Forced test crash from NativePHP FirebaseCrashlytics::crash()"

            Handler(Looper.getMainLooper()).post {
                throw RuntimeException(reason)
            }

            return BridgeResponse.success(mapOf("crashing" to true))
        }
    }

    /**
     * Parameters:
     * - enabled: Boolean
     *
     * Returns:
     * - enabled: Boolean
     */
    class SetCollectionEnabled(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val enabled = parameters["enabled"] as? Boolean ?: true

            FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(enabled)

            return BridgeResponse.success(mapOf("enabled" to enabled))
        }
    }

    /**
     * Returns:
     * - enabled: Boolean
     */
    class IsCollectionEnabled(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return BridgeResponse.success(mapOf(
                "enabled" to FirebaseCrashlytics.getInstance().isCrashlyticsCollectionEnabled
            ))
        }
    }

    /**
     * Blocks (with a timeout) on the async Crashlytics Task so PHP receives a
     * synchronous boolean result.
     *
     * Returns:
     * - hasUnsentReports: Boolean
     */
    class CheckForUnsentReports(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return try {
                val hasReports = Tasks.await(
                    FirebaseCrashlytics.getInstance().checkForUnsentReports(),
                    AWAIT_TIMEOUT_SECONDS,
                    TimeUnit.SECONDS
                )

                BridgeResponse.success(mapOf("hasUnsentReports" to hasReports))
            } catch (e: TimeoutException) {
                BridgeResponse.error("TIMEOUT", "Timed out checking for unsent reports")
            } catch (e: Exception) {
                BridgeResponse.error("CHECK_FAILED", e.message ?: "Failed to check for unsent reports")
            }
        }
    }

    /**
     * Returns:
     * - sent: Boolean
     */
    class SendUnsentReports(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            FirebaseCrashlytics.getInstance().sendUnsentReports()

            return BridgeResponse.success(mapOf("sent" to true))
        }
    }

    /**
     * Returns:
     * - deleted: Boolean
     */
    class DeleteUnsentReports(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            FirebaseCrashlytics.getInstance().deleteUnsentReports()

            return BridgeResponse.success(mapOf("deleted" to true))
        }
    }

    /**
     * Returns:
     * - didCrash: Boolean
     */
    class DidCrashOnPreviousExecution(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return BridgeResponse.success(mapOf(
                "didCrash" to FirebaseCrashlytics.getInstance().didCrashOnPreviousExecution()
            ))
        }
    }

    /**
     * Lightweight throwable used to preserve the PHP exception's class name
     * and message when forwarding it to Crashlytics as a recorded exception.
     * It is never thrown - only passed to recordException() for its class
     * name/message/synthetic stack trace.
     */
    private class PhpThrowable(className: String, message: String) : Throwable("$className: $message")
}

File: resources/ios/FirebaseCrashlyticsFunctions.swift
import FirebaseCrashlytics
import Foundation

enum FirebaseCrashlyticsFunctions {

    /// How long we're willing to wait on the async checkForUnsentReports callback.
    private static let awaitTimeoutSeconds: TimeInterval = 8.0

    /// Records a caught exception with Crashlytics as a (non-)fatal issue.
    ///
    /// - Parameters:
    ///   - message: String - Exception message
    ///   - className: String? - Fully qualified PHP exception class name, used as the NSError domain
    ///   - stackTrace: String? - Rendered PHP stack trace, logged as a breadcrumb before recording
    ///   - fatal: Bool - Whether to flag this issue as fatal-style via a custom key
    ///
    /// - Returns: `recorded: Bool`
    class RecordError: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let message = parameters["message"] as? String else {
                return BridgeResponse.error(code: "INVALID_PARAMETERS", message: "message is required")
            }
            let className = parameters["className"] as? String ?? "PHP.Throwable"
            let stackTrace = parameters["stackTrace"] as? String
            let fatal = parameters["fatal"] as? Bool ?? false

            let crashlytics = Crashlytics.crashlytics()

            if fatal {
                crashlytics.setCustomValue(true, forKey: "fatal_error")
            }
            if let stackTrace = stackTrace {
                crashlytics.log("PHP stack trace:\n\(stackTrace)")
            }

            let error = NSError(
                domain: className,
                code: 0,
                userInfo: [NSLocalizedDescriptionKey: message]
            )
            crashlytics.record(error: error)

            return BridgeResponse.success(data: ["recorded": true])
        }
    }

    /// - Parameters:
    ///   - message: String - Breadcrumb message to attach to the next report
    ///
    /// - Returns: `logged: Bool`
    class Log: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let message = parameters["message"] as? String else {
                return BridgeResponse.error(code: "INVALID_PARAMETERS", message: "message is required")
            }

            Crashlytics.crashlytics().log(message)

            return BridgeResponse.success(data: ["logged": true])
        }
    }

    /// - Parameters:
    ///   - identifier: String - User identifier to attach to subsequent reports
    ///
    /// - Returns: `set: Bool`
    class SetUserId: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let identifier = parameters["identifier"] as? String else {
                return BridgeResponse.error(code: "INVALID_PARAMETERS", message: "identifier is required")
            }

            Crashlytics.crashlytics().setUserID(identifier)

            return BridgeResponse.success(data: ["set": true])
        }
    }

    /// - Parameters:
    ///   - key: String - Custom key name
    ///   - value: String|Bool|Number - Custom key value
    ///
    /// - Returns: `set: Bool`
    class SetCustomKey: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let key = parameters["key"] as? String else {
                return BridgeResponse.error(code: "INVALID_PARAMETERS", message: "key is required")
            }
            guard let value = parameters["value"] else {
                return BridgeResponse.error(code: "INVALID_PARAMETERS", message: "value is required")
            }

            Crashlytics.crashlytics().setCustomValue(value, forKey: key)

            return BridgeResponse.success(data: ["set": true])
        }
    }

    /// Forces an immediate, unrecoverable crash - used to verify a Crashlytics
    /// integration end-to-end. `fatalError` traps unconditionally and cannot be
    /// intercepted by any `try`/`catch`, guaranteeing a real crash.
    ///
    /// - Parameters:
    ///   - reason: String? - Optional crash message
    class Crash: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let reason = parameters["reason"] as? String
                ?? "Forced test crash from NativePHP FirebaseCrashlytics::crash()"

            fatalError(reason)
        }
    }

    /// - Parameters:
    ///   - enabled: Bool
    ///
    /// - Returns: `enabled: Bool`
    class SetCollectionEnabled: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let enabled = parameters["enabled"] as? Bool ?? true

            Crashlytics.crashlytics().isCrashlyticsCollectionEnabled = enabled

            return BridgeResponse.success(data: ["enabled": enabled])
        }
    }

    /// - Returns: `enabled: Bool`
    class IsCollectionEnabled: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            return BridgeResponse.success(data: [
                "enabled": Crashlytics.crashlytics().isCrashlyticsCollectionEnabled
            ])
        }
    }

    /// Blocks (with a timeout) on the async completion handler so PHP receives
    /// a synchronous boolean result.
    ///
    /// - Returns: `hasUnsentReports: Bool`
    class CheckForUnsentReports: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let semaphore = DispatchSemaphore(value: 0)
            var hasReports = false

            Crashlytics.crashlytics().checkForUnsentReports { result in
                hasReports = result
                semaphore.signal()
            }

            let result = semaphore.wait(timeout: .now() + FirebaseCrashlyticsFunctions.awaitTimeoutSeconds)

            if result == .timedOut {
                return BridgeResponse.error(code: "TIMEOUT", message: "Timed out checking for unsent reports")
            }

            return BridgeResponse.success(data: ["hasUnsentReports": hasReports])
        }
    }

    /// - Returns: `sent: Bool`
    class SendUnsentReports: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            Crashlytics.crashlytics().sendUnsentReports()

            return BridgeResponse.success(data: ["sent": true])
        }
    }

    /// - Returns: `deleted: Bool`
    class DeleteUnsentReports: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            Crashlytics.crashlytics().deleteUnsentReports()

            return BridgeResponse.success(data: ["deleted": true])
        }
    }

    /// - Returns: `didCrash: Bool`
    class DidCrashOnPreviousExecution: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            return BridgeResponse.success(data: [
                "didCrash": Crashlytics.crashlytics().didCrashDuringPreviousExecution()
            ])
        }
    }
}
